File: 007_ReverseInteger.c



//8.42%
int reverse(int x) {
    //Check boundary conditions
	int valX = x;
	//Consider INT_MAX && INT_MIN
	#define INT_MAX 0x7fffffff
	#define INT_MIN 0x80000000
	if(x == INT_MIN)
		return 0;
	if(x < 0)
		valX = -x;
	if(valX < 10)
		return x;
	
	//Main routine
	long long rstX = 0;
	while(valX)
	{
		int unitX = valX % 10;
		rstX *= 10;
		rstX += unitX;
		valX /= 10;
	}
	
	if(rstX <= INT_MAX)
	{
		if(x > 0)
			return (int)rstX;
		else
			return -(int)rstX;
	}
	else
		return 0;
}

//24.6%
int reverse001(int x) {
    //Check boundary conditions
	int valX = x;
	//Consider INT_MAX && INT_MIN
	unsigned intMax = (unsigned)(~0) >> 1;
	unsigned intMin = intMax + 1;
	if(x == intMin)
		return 0;
	if(x < 0)
		valX = -x;
	if(valX < 10)
		return x;
	
	//Main routine
	long long unsigned rstX = 0;
	while(valX)
	{
		int unitX = valX % 10;
		rstX *= 10;
		rstX += unitX;
		valX /= 10;
	}
	

	if(rstX <= intMax)
	{
		if(x > 0)
			return (int)rstX;
		else
			return -(int)rstX;
	}
	else if(rstX == intMin && x < 0)
		return (int)rstX;
	else
		return 0;
}

File: 007_ReverseInteger.py

#0.53%
class Solution(object):
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        #Check boundary conditions
        intMax = 0x7fffffff
        intMin = 0x80000000
        if -10 < x and x < 10:
            return x
        if x == intMin:
            return 0
            
        #Main routine
        flag = True
        if x < 0:
            flag = False
            x = -x
        rstVal = 0
        while x != 0:
            unitVal = x % 10
            rstVal = rstVal * 10 + unitVal
            x = x / 10
        
        #Check Integer overflow
        if rstVal > intMax:
            return 0
        if not flag:
            rstVal = -rstVal
            
        return rstVal
            